[19.0] [MIG] queue_job: migrate + tests

tishmen opened the pull request:
Scope
- queue_job, test_queue_job only

Summary
- Migrate to Odoo 19 using the working feature branch code.
- Tests create their own users (no demo data), context handling aligned with 19.0.
- Asset paths are relative under web.assets_backend.

Pre-commit
- Verified locally; no additional changes applied here to keep parity with the feature branch.

Tests
- Command:
  ./odoo/odoo-bin -c odoo.conf -d odoo_queue_pr_queue_job_base_001 --init queue_job,test_queue_job --test-tags="/queue_job,/test_queue_job" --stop-after-init
- Result: all tests green locally.

sbidoul commented:
/ocabot migration queue_job

sbidoul commented:
/ocabot migration test_queue_job

hoangtrann commented on queue_job/controllers/main.py:
```diff
@@ -11,7 +11,7 @@
 from werkzeug.exceptions import BadRequest, Forbidden
 
-from odoo import SUPERUSER_ID, _, api, http
+from odoo import SUPERUSER_ID, api, http
```
I believe `from odoo import _` is still working and still widely used in odoo code-base

tishmen replied:
I switched to using http.request.env.(...) in the controller to comply with the repo’s linting policy (pylint-odoo’s prefer-env-translation). Tying translations to the current request/env guarantees the user’s language is used in HTTP contexts. After that change, the global “_” import became unused, so Ruff removed it (unused-import). No behavior change was intended, this is purely to satisfy the configured linters. If the project prefers the global “_”, I can adjust the linter rule accordingly.

sbidoul replied:
I also had my doubts about this change for translation, as it makes the code much more verbose. But it improves performance a tiny bit. So I'm getting used to it.

hoangtrann replied:
yeah, I just cannot wrap my head around `self.env._` for translation... and it isn't aligned with the odoo code-base so it would also looks strange to many

hoangtrann commented on queue_job/controllers/main.py:
```diff
@@ -280,7 +280,7 @@ def _create_graph_test_jobs(
                         max_retries=max_retries,
                         channel=channel,
-                        description="%s #%d" % (description, current_count),
+                        description=f"{description} #{current_count}",
```
this should be avoided in a migration PR, unless maintainer(s) say otherwise

tishmen replied:
Thanks for pointing it out. That formatting change came from the pre-commit linters (Ruff’s pyupgrade UP008), which auto-converts percent-formatting to f-strings based on the config (see queue/.ruff.toml and .pre-commit-config.yaml). No behavior was intended to change, it’s purely lint-driven. If you prefer to keep migration diffs minimal, I can revert this change.

sbidoul replied:
Switching to f-strings improves readability. So lets keep that.

hoangtrann replied:
@sbidoul should we convert everything to f-strings then? I believe it should be either or. We should also preserve the format though, so `"#%d"` should be `f"{:d}"`

tishmen replied:
@sbidoul @hoangtrann  I’ll standardize on f-strings across the codebase, with a couple of exceptions for best practices:

- Logging calls remain parameterized (e.g., _logger.info("… %s", x)) rather than f-strings, to preserve lazy formatting and match Odoo’s logging style.
- Translatable strings and SQL placeholders also stay as-is.
- Everywhere else (exceptions, non-i18n messages, test strings), I’ll switch to f-strings and preserve formatting specs (e.g., “#%d” → f“{value:d}”, “%.2f” → f“{value:.2f}”).
- I’ll also extract f-string error messages into variables before raising (e.g., msg = f"..."; raise ValueError(msg)). This improves readability and makes it easier to handle or log the message separately if needed.

hoangtrann commented on queue_job/fields.py:
```diff
@@ -66,7 +67,7 @@ def convert_to_cache(self, value, record, validate=True):
     def convert_to_record(self, value, record):
         default = self._base_type_default_json(record.env)
         value = value or default
-        if not isinstance(value, (str | bytes | bytearray)):
+        if not isinstance(value, (str, bytes, bytearray)):
```
remove this change, `isinstance(value, (str | bytes | bytearray))` is the more modern and is accurate

tishmen replied:
You’re right. The tuple form is fully supported and widely used in Odoo 19’s minimum Python (3.10). Both styles are valid in 3.10+: tuple (str, bytes, bytearray) and PEP 604 unions (str | bytes | bytearray). Core code often uses the tuple form, while some addons use the union operator. I’m fine keeping the tuple here for consistency with core; functionally they’re equivalent.

hoangtrann commented on queue_job/job.py:
```diff
@@ -505,7 +505,7 @@ def perform(self):
                 # http://blog.ianbicking.org/2007/09/12/re-raising-exceptions/
                 new_exc = FailedJobError(
-                    "Max. retries (%d) reached: %s" % (self.max_retries, value or type_)
+                    f"Max. retries ({self.max_retries}) reached: {value or type_}"
```
should be avoid in migration PR

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404065707

hoangtrann commented on queue_job/job.py:
```diff
@@ -813,7 +813,7 @@ def set_failed(self, **kw):
 
     def __repr__(self):
-        return "<Job %s, priority:%d>" % (self.uuid, self.priority)
+        return f"<Job {self.uuid}, priority:{self.priority}>"
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404065707

hoangtrann commented on queue_job/job.py:
```diff
@@ -828,7 +828,7 @@ def _get_retry_seconds(self, seconds=None):
         elif not seconds:
             seconds = RETRY_INTERVAL
-        if isinstance(seconds, (list | tuple)):
+        if isinstance(seconds, (list, tuple)):
```
same above

tishmen replied:
Answered at: https://github.com/OCA/queue/pull/840#discussion_r2401927895

hoangtrann commented on queue_job/jobrunner/channels.py:
```diff
@@ -455,12 +455,9 @@ def get_subchannel_by_name(self, subchannel_name):
-            len(self._queue),
-            len(self._running),
-            len(self._failed),
+        return (
+            f"{self.fullname}(C:{capacity},Q:{len(self._queue)},"
+            f"R:{len(self._running)},F:{len(self._failed)})"
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404065707

hoangtrann commented on queue_job/models/base.py:
```diff
@@ -262,9 +262,26 @@ def _job_prepare_context_before_enqueue(self):
         method.origin = origin
-        # propagate decorators from origin to method, and apply api decorator
-        wrapped = api.propagate(origin, method)
+        # carry over wrapper attributes (name, doc, etc.)
+        wrapped = functools.update_wrapper(method, origin)
+        # propagate common decorator metadata used by the framework
+        for attr in (
+            "_constrains",
+            "_depends",
+            "_onchange",
+            "_ondelete",
+            "_api_model",
+            "_api_private",
+        ):
+            if hasattr(origin, attr):
+                setattr(wrapped, attr, getattr(origin, attr))
```
have you considered just reimplemented the same old `propagate` method instead?

tishmen replied:
In 19.0, api.propagate was removed as part of the API/ORM refactor (api now re‑exports from odoo.orm.*) and upstream uses functools.update_wrapper directly. To remain compatible and avoid reintroducing a removed internal API, we mirror the upstream approach: update_wrapper for name/doc and explicit propagation of the few decorator flags that Odoo actually reads (_constrains/_depends/_onchange/_ondelete/_api_model/_api_private). This keeps the behavior identical while minimizing reliance on non‑public helpers. If you prefer a local helper to reduce duplication, I can wrap this logic in a small internal function, but I’d avoid reviving the “propagate” API name.

tishmen replied:
@sbidoul can we have your input?

hoangtrann replied:
I don't know what the best to do here, but hard coding a list of apis here also don't work as a proper propagation. because if there is another api being added through customization, it will be skipped by this implementation I think :thinking:

sbidoul replied:
It looks like this `_patch_method` was introduced in `queue_job` in 17.0, as a copy of the method that was in upstream `base` in Odoo 16.0. I think this was not a good solution in the first place. I'm not sure what to do, but at least it should be renamed.

sbidoul replied:
`_patch_method` is not actually used in `queue_job`. It is only mentioned in the comment of `_patch_job_auto_delay` . I'm inclined to remove `_patch_method` from `queue_job`.

So I'm inclined to remove it. If it is necessary to reintroduce it, that should happen in another standalone module. @guewen do you have an opinion on this?

tishmen replied:
@sbidoul @guewen @hoangtrann Addressed in: https://github.com/OCA/queue/pull/840/commits/038c334038d5b3c479763cdddb702cf5826489aa. 

Please review.

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,49 @@ class QueueJob(models.Model):
-                "CREATE INDEX queue_job_identity_key_state_partial_index "
-                "ON queue_job (identity_key) WHERE state in ('pending', "
-                "'enqueued', 'wait_dependencies') AND identity_key IS NOT NULL;"
+            create_index(
+                cr,
+                index_1,
+                "queue_job",
+                ["identity_key"],
+                where=(
+                    "state in ('pending','enqueued','wait_dependencies') "
+                    "AND identity_key IS NOT NULL"
+                ),
+                comment=("Queue Job: partial index for identity_key on active states"),
```
I would say this should be avoided in migration PR, just replaced `self._cr` with `self.env.cr` is good enough for mig

tishmen replied:
I replaced the raw CREATE INDEX with odoo.tools.sql.create_index to align with Odoo 19’s recommended API (see odoo/odoo/tools/sql.py) and to satisfy pylint-odoo’s sql-injection rule. Core addons (e.g., project, phone_validation) use the same helper. If you prefer a strictly minimal migration, I can switch to self.env.cr and keep the raw SQL, but that will either trigger the linter or need a local disable.

tishmen replied:
@sbidoul can we have your input?

sbidoul replied:
Let's use `create_index`, yes. `if not index_exists` is redundant, though, as `create_index` does it by itself.

tishmen replied:
Resolved in: https://github.com/OCA/queue/pull/840/commits/9ea78c5ed72d2b57287f084b49b9f2ae49e8b763

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,49 @@ class QueueJob(models.Model):
+                index_2,
+                "queue_job",
+                ["channel", "date_done", "date_created"],
+                comment="Queue Job: index to accelerate autovacuum",
```
same above

tishmen replied:
Answered at: https://github.com/OCA/queue/pull/840#discussion_r2404121212

sbidoul replied:
Remove `if not index_exists`

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -247,11 +259,12 @@ def write(self, vals):
-                raise exceptions.AccessError(
-                    _("Not allowed to change field(s): {}").format(
-                        write_on_protected_fields
-                    )
+                # use env translation and lazy formatting (args to _)
+                msg = self.env._(
+                    "Not allowed to change field(s): %s",
+                    ", ".join(write_on_protected_fields),
                 )
+                raise exceptions.AccessError(msg)
```
I think we shouldn't change this as it still working

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

sbidoul replied:
Having the message built outside the raise statement is good practice, as it makes stack traces more readable. :+1:

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -279,7 +292,7 @@ def open_related_action(self):
         action = job.related_action()
         if action is None:
-            raise exceptions.UserError(_("No action available for this job"))
+            raise exceptions.UserError(self.env._("No action available for this job"))
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -292,7 +305,7 @@ def open_graph_jobs(self):
         action.update(
             {
-                "name": _("Jobs for graph %s") % (self.graph_uuid),
+                "name": self.env._("Jobs for graph %s", self.graph_uuid),
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -324,12 +337,12 @@ def _change_job_state(self, state, result=None):
 
     def button_done(self):
-        result = _("Manually set to done by {}").format(self.env.user.name)
+        result = self.env._("Manually set to done by %s", self.env.user.name)
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -324,12 +337,12 @@ def _change_job_state(self, state, result=None):
 
     def button_cancelled(self):
-        result = _("Cancelled by {}").format(self.env.user.name)
+        result = self.env._("Cancelled by %s", self.env.user.name)
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -370,7 +383,7 @@ def _message_failed_job(self):
         """
         self.ensure_one()
-        return _(
+        return self.env._(
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -388,25 +401,36 @@ def autovacuum(self):
+                        if not config["test_enable"]:
+                            self.env.cr.commit()  # pylint: disable=E8102
+                    else:
+                        break
```
this should be avoid in migration pr

tishmen replied:
The loop and batching here are to satisfy pylint-odoo’s no-search-all rule (unbounded search([])), which is enabled in both our optional and mandatory configs. We also delete in chunks with a stable order to match the autovacuum index and commit between batches to avoid long transactions; the commit is guarded for tests and has a targeted disable for the invalid-commit rule. This keeps behavior the same while conforming to the linter and avoids performance issues on large tables. If you prefer to keep the migration strictly minimal, I can revert this hunk and add a local disable for no-search-all (and invalid-commit), but this version aligns with our 19.0 lint and index usage.

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -425,7 +449,7 @@ def related_action_open_record(self):
             return None
         action = {
-            "name": _("Related Record"),
+            "name": self.env._("Related Record"),
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job.py:
```diff
@@ -435,7 +459,7 @@ def related_action_open_record(self):
             action.update(
                 {
-                    "name": _("Related Records"),
+                    "name": self.env._("Related Records"),
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_channel.py:
```diff
@@ -26,9 +26,11 @@ class QueueJobChannel(models.Model):
-    _sql_constraints = [
-        ("name_uniq", "unique(complete_name)", "Channel complete name must be unique")
-    ]
+    # Odoo 19: _sql_constraints removed. Use models.Constraint instead.
```
comment is unneccessary

tishmen replied:
Ok, will remove the unnecessary comment.

hoangtrann commented on queue_job/models/queue_job_channel.py:
```diff
@@ -45,7 +47,7 @@ def _compute_complete_name(self):
         for record in self:
             if record.name != "root" and not record.parent_id:
-                raise exceptions.ValidationError(_("Parent channel required."))
+                raise exceptions.ValidationError(self.env._("Parent channel required."))
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_channel.py:
```diff
@@ -79,11 +81,15 @@ def write(self, values):
                 and ("name" in values or "parent_id" in values)
             ):
-                raise exceptions.UserError(_("Cannot change the root channel"))
+                raise exceptions.UserError(self.env._("Cannot change the root channel"))
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_channel.py:
```diff
@@ -79,11 +81,15 @@ def write(self, values):
+                raise exceptions.UserError(self.env._("Cannot change the root channel"))
         return super().write(values)
 
     def unlink(self):
+        # Do not raise here to comply with lint; guard in ondelete instead.
+        return super().unlink()
```
then remove it completely?

tishmen replied:
Thanks. I just double‑checked: removing the unlink override keeps all queue tests green. The root deletion guard is enforced in @api.ondelete (_check_not_root_ondelete), so the override was a no‑op. I’ll delete it in this PR as suggested.

hoangtrann commented on queue_job/models/queue_job_channel.py:
```diff
@@ -79,11 +81,15 @@ def write(self, values):
             if channel.name == "root":
-                raise exceptions.UserError(_("Cannot remove the root channel"))
-        return super().unlink()
+                raise exceptions.UserError(self.env._("Cannot remove the root channel"))
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_function.py:
```diff
@@ -91,14 +91,16 @@ def _compute_name(self):
         groups = regex_job_function_name.match(self.name)
         if not groups:
-            raise exceptions.UserError(_("Invalid job function: {}").format(self.name))
+            raise exceptions.UserError(
+                self.env._("Invalid job function: %s", self.name)
+            )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_function.py:
```diff
@@ -91,14 +91,16 @@ def _compute_name(self):
         )
         if not model:
-            raise exceptions.UserError(_("Model {} not found").format(model_name))
+            raise exceptions.UserError(self.env._("Model %s not found", model_name))
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_function.py:
```diff
@@ -187,12 +189,13 @@ def job_config(self, name):
-            "Unexpected format of Retry Pattern for {}.\n"
+        return self.env._(
+            "Unexpected format of Retry Pattern for %s.\n"
             "Example of valid formats:\n"
             "{{1: 300, 5: 600, 10: 1200, 15: 3000}}\n"
-            "{{1: (1, 10), 5: (11, 20), 10: (21, 30), 15: (100, 300)}}"
-        ).format(self.name)
+            "{{1: (1, 10), 5: (11, 20), 10: (21, 30), 15: (100, 300)}}",
+            self.name,
+        )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_function.py:
```diff
@@ -211,20 +214,21 @@ def _check_retry_pattern(self):
 
     def _retry_value_type_check(self, value):
-        if isinstance(value, (tuple | list)):
+        if isinstance(value, (tuple, list)):
```
same above

tishmen replied:
Answered at: https://github.com/OCA/queue/pull/840#discussion_r2401927895

hoangtrann commented on queue_job/models/queue_job_function.py:
```diff
@@ -211,20 +214,21 @@ def _check_retry_pattern(self):
     def _related_action_format_error_message(self):
-        return _(
-            "Unexpected format of Related Action for {}.\n"
+        return self.env._(
+            "Unexpected format of Related Action for %s.\n"
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/models/queue_job_function.py:
```diff
@@ -211,20 +214,21 @@ def _check_retry_pattern(self):
+        return self.env._(
+            "Unexpected format of Related Action for %s.\n"
             "Example of valid format:\n"
             '{{"enable": True, "func_name": "related_action_foo",'
-            ' "kwargs" {{"limit": 10}}}}'
-        ).format(self.name)
+            ' "kwargs" {{"limit": 10}}}}',
+            self.name,
+        )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404063922

hoangtrann commented on queue_job/security/security.xml:
```diff
@@ -1,16 +1,23 @@
 <?xml version="1.0" encoding="utf-8" ?>
 <odoo>
     <data>
+        <!-- New: Odoo 19 groups are classified by res.groups.privilege -->
```
remove comment

tishmen replied:
Ok, will remove the unnecessary comment.

hoangtrann commented on queue_job/security/security.xml:
```diff
@@ -1,16 +1,23 @@
+        <!-- New: Odoo 19 groups are classified by res.groups.privilege -->
+        <record model="res.groups.privilege" id="privilege_queue_job">
+            <field name="name">Job Queue</field>
+            <field name="sequence">20</field>
```
sequence should be a bit higher, 50 maybe

tishmen replied:
Thanks, I checked core for reference points.

auth_oauth.menu_oauth_providers has sequence="30" under Settings › Users (base.menu_users). To keep “Job Queue” from crowding the business privileges, I’ll bump its privilege sequence to 50.

hoangtrann commented on queue_job/security/security.xml:
```diff
@@ -1,16 +1,23 @@
-            <field name="category_id" ref="module_category_queue_job" />
+            <!-- Odoo 19: use privilege_id instead of deprecated category_id -->
+            <field name="privilege_id" ref="privilege_queue_job" />
+            <!-- Odoo 19: res.groups uses user_ids M2M and Command API -->
```
no need comment

tishmen replied:
Ok, will remove the unnecessary comment.

hoangtrann commented on queue_job/tests/common.py:
```diff
@@ -335,7 +334,8 @@ def search_created(self):
 
     def search_all(self):
-        return self.env["queue.job"].search([])
+        # Avoid unbounded search([]) for lint compliance
+        return self.env["queue.job"].search([("id", "!=", 0)])
```
I suspect this is the case, what's the lint error?

tishmen replied:
This change is linter-driven. The pylint-odoo config enables the “no-search-all” rule (flags unbounded search([]) calls):

queue/.pylintrc:113
queue/.pylintrc-mandatory:105
To satisfy that rule while keeping the same semantics in tests, I replaced search([]) with search([("id", "!=", 0)]). No functional behavior change is intended; it’s purely to comply with the configured linter.

If you prefer to keep search([]) in test code, I can revert and add a targeted disable (e.g., “# pylint: disable=no-search-all”) or we can relax the rule for tests.

sbidoul replied:
`search([])` is much much more readable, let's silence the linter locally if it is the only place. I'm not quite convinced about this lint rule.

tishmen replied:
Thanks, agreed, readability wins here.

I’ll revert to the original:
return self.env["queue.job"].search([]) # pylint: disable=no-search-all
and keep the disable scoped to this test helper so pre-commit stays green without changing behavior. I’ll also simplify the part you pointed out in https://github.com/OCA/queue/pull/840#discussion_r2401984382 as well.

hoangtrann commented on queue_job/tests/common.py:
```diff
@@ -414,15 +414,21 @@ def test_export(self):
         yield delayable_cls, delayable
 
 
-class OdooDocTestCase(doctest.DocTestCase, _TestCase, MetaCase("DummyCase", (), {})):
+class OdooDocTestCase(doctest.DocTestCase, _TestCase):
```
I'm not sure what's the usage of MetaCase, but could you explain why removing this still preserve the behavior?

tishmen replied:
MetaCase doesn’t exist in Odoo 19 and isn’t used anywhere in core. The behavior MetaCase used to provide (class-level cleanup/compat) is now built into Odoo’s TestCase (odoo.tests.case.TestCase). Our doctest harness remains intact with doctest.DocTestCase + _TestCase, so dropping MetaCase keeps the tests working while aligning with Odoo 19’s test framework.

tishmen replied:
@sbidoul can we have your input on this?

tishmen replied:
@hoangtrann I'm marking this one as resolved.

sbidoul replied:
It looks like doctests are not executed, but we can handle that in a followup. I tested manually that they still pass.

hoangtrann commented on queue_job/tests/common.py:
```diff
@@ -414,15 +414,21 @@ def test_export(self):
-        self, doctest, optionflags=0, setUp=None, tearDown=None, checker=None, seq=0
+        self,
+        doctest,
+        optionflags=doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE,
```
why change from 0 to `doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE`?

tishmen replied:
All set. I reran the full test suite on a fresh database with the doctest flags set back to 0, and everything passes. I'll revert

hoangtrann commented on queue_job/tests/test_json_field.py:
```diff
@@ -14,8 +14,24 @@
 
 
 class TestJson(common.TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
+            {
+                "name": "Demo User (Queue)",
+                "login": "queue_demo_user",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
+            }
+        )
```
why adding this now in the mig?

tishmen replied:
The added user in setUpClass is to remove reliance on demo data, which isn’t loaded by default in OCA’s 19.0 test environment.  (see queue/.github/workflows/test.yml using oca_init_test_database). Odoo itself also doesn’t load demo unless explicitly requested. The previous tests referenced base.user_demo, which fails when demo isn’t loaded. Creating a minimal user in the test setup makes the tests self-contained and reliable across environments. This aligns with OCA testing practice to avoid demo dependencies. If you prefer, I can gate it to use base.user_demo when present and only create the user as a fallback.

hoangtrann replied:
we should keep it then :+1:

PieterPaulussen replied:
Aren't unittests always run on a database with demo data present?

hoangtrann commented on queue_job/views/queue_job_views.xml:
```diff
@@ -256,68 +253,57 @@
+                <separator />
+                <!-- Odoo 19: no <group expand> in search; declare group-by filters inline -->
                 <filter
-                    name="last_24_hours"
-                    string="Last 24 hours"
-                    domain="[('date_created', '&gt;=', (context_today() - datetime.timedelta(days=1)).strftime('%Y-%m-%d'))]"
```
could we do something similar as done here? https://github.com/odoo/odoo/pull/228026

tishmen replied:
Good call, using the dynamic domain macros is much cleaner. I’ll switch the filters to the core-friendly form (e.g., 'now -1d', 'today', '=week_start +1w'). I had been looking for a core example earlier and missed the timesheet PR; thanks for the pointer.

hoangtrann commented on queue_job/views/queue_job_views.xml:
```diff
@@ -256,68 +253,57 @@
+                    context="{'group_by': 'channel'}"
                 />
                 <filter
-                    name="last_7_days"
-                    string="Last 7 days"
-                    domain="[('date_created', '&gt;=', (context_today() - datetime.timedelta(days=7)).strftime('%Y-%m-%d'))]"
```
same above

tishmen replied:
Answered at: https://github.com/OCA/queue/pull/840#discussion_r2404147675

hoangtrann commented on queue_job/views/queue_job_views.xml:
```diff
@@ -256,68 +253,57 @@
+                    context="{'group_by': 'job_function_id'}"
                 />
                 <filter
-                    name="last_30_days"
-                    string="Last 30 days"
-                    domain="[('date_created', '&gt;=', (context_today() - datetime.timedelta(days=30)).strftime('%Y-%m-%d'))]"
```
same above

tishmen replied:
Answered at: https://github.com/OCA/queue/pull/840#discussion_r2404147675

hoangtrann commented on test_queue_job/tests/test_job.py:
```diff
@@ -32,6 +32,22 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
+            {
+                "name": "Demo User (Queue)",
+                "login": "queue_demo_user_3",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
+            }
+        )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404078382

hoangtrann commented on test_queue_job/tests/test_job.py:
```diff
@@ -387,6 +403,22 @@ def test_job_identity_key_func_exact(self):
 
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
+            {
+                "name": "Demo User (Queue)",
+                "login": "queue_demo_user_4",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
+            }
+        )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404078382

hoangtrann commented on test_queue_job/tests/test_job.py:
```diff
@@ -511,6 +543,22 @@ def test_store_env_su_sudo(self):
 class TestJobModel(JobCommonCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
+            {
+                "name": "Demo User (Queue)",
+                "login": "queue_demo_user_5",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
+            }
+        )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404078382

hoangtrann commented on test_queue_job/tests/test_json_field.py:
```diff
@@ -11,17 +11,33 @@
 class TestJsonField(common.TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
+            {
+                "name": "Demo User (Queue)",
+                "login": "queue_demo_user_2",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
+            }
+        )
```
same above

tishmen replied:
Answered here: https://github.com/OCA/queue/pull/840#discussion_r2404078382

hoangtrann reviewed (changes requested):
appreciate all the work you done here! please review the comments since I notice many weird comments and many changes should be in another PR instead of a migration one

tishmen commented:
@hoangtrann quick update:

- We’re down to just **3 open comments** on this PR; the rest are addressed and pushed.
- When you have a moment, a fresh check would be super helpful.

I’d also really appreciate it if you could take a look at **PR #842**—specifically the **tests**:  

You can run the suite locally with (use any throwaway DB name in place of `<DB_NAME>`):

./odoo/odoo-bin -c odoo.conf -d <DB_NAME> \
  --init queue_job,queue_job_batch,queue_job_cron,queue_job_cron_jobrunner,queue_job_subscribe,test_queue_job,test_queue_job_batch \
  --test-tags="/queue_job,/queue_job_batch,/queue_job_cron,/queue_job_cron_jobrunner,/queue_job_subscribe,/test_queue_job,/test_queue_job_batch" \
  --stop-after-init

Thanks again for the **thorough and detailed review**—really grateful for your contributions and feedback 🙏

sbidoul commented on queue_job/jobrunner/__init__.py:
```diff
@@ -10,17 +10,46 @@
+    # from flat odoo.conf options so the runner works without server_environment.
+    queue_job_config = {}
+
+# Merge flat odoo.conf options as a fallback (applies regardless of whether
```
It looks like this merging logic changes option parsing behaviour? In that case it's better to that in a separate PR.

hoangtrann replied:
@sbidoul since `config.misc` is removed in odoo 19, we might need to drop it here entirely and open another pr depends on this one to introduce back the behavior for a clean migration pr

sbidoul replied:
What I mean is that we should not interoduce a new merging logic that was not there before. That would be a new feature, which is not to be in a migration PR.

I think something like this should work in case `server_environment` is not available:

```
from configparser import ConfigParser

queue_job_config = {}
if config["config"]:
    cp = ConfigParser(interpolation=None)
    cp.read(config["config"])  # read the odoo config file
    if cp.has_section("queue_job"):
        queue_job_config = cp["queue_job"]
```

tishmen replied:
Resolved in: https://github.com/OCA/queue/pull/840/commits/231705d0758e1b14246972cdab9d664eac8c262f

sbidoul commented on queue_job/tests/common.py:
```diff
@@ -305,11 +304,12 @@ def _format_job_call(self, call):
+            f"{call.method.__name__}("
+            f"{', '.join(method_all_args)}) "
+            f"with properties ("
+            f"{', '.join(f'{key}={value}' for key, value in call.properties.items())})"
```
Nit: sometimes f-strings do degrade readability. Here intermediate variables would help.

tishmen replied:
Resolved in: https://github.com/OCA/queue/pull/840/commits/8dd4af0c541683882a5f1970e52c3f18091a8165

sbidoul commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,49 @@ class QueueJob(models.Model):
     worker_pid = fields.Integer(readonly=True)
 
     def init(self):
+        # Odoo 19: self._cr deprecated; use self.env.cr.
+        # Prefer tools.sql helpers for idempotent DDL.
```
These Odoo 19 comments will not help much we we look at future versions. They were helpful for code review, but now it's better to remove them.

tishmen replied:
Resolved in [#d27f83d8c4b3a0a2a41c7cb2c71dc555db20bc0e](https://github.com/OCA/queue/pull/840/commits/d27f83d8c4b3a0a2a41c7cb2c71dc555db20bc0e)

hoangtrann reviewed (commented):
there are still some comments stating odoo 19 scattered across the changes which I thing we should removed.

for anyone comes after these are removed, I think it's good to just summarized the major changes to the description of the pr which can also works as a good resource for reference in other mig pr

tishmen commented:
> there are still some comments stating odoo 19 scattered across the changes which I thing we should removed.
> 
> for anyone comes after these are removed, I think it's good to just summarized the major changes to the description of the pr which can also works as a good resource for reference in other mig pr

Resolved in: https://github.com/OCA/queue/pull/840/commits/d27f83d8c4b3a0a2a41c7cb2c71dc555db20bc0e

hoangtrann reviewed (approved):
LGTM :+1:

richard-willdooit commented:
From my checks and tests, this is all ready to go.  Can it be merged?

@sbidoul

PieterPaulussen commented:
@tishmen have you tested this version with the preforkserver configuration (using `--workers=4`)? Jobs are enqueued, but never started.
The threaded server setup does work.

Additionally, there is an error due to the upstream config refactoring in this method:
https://github.com/tishmen/queue/blob/pr/queue_job_base/queue_job/jobrunner/runner.py#L474-L479. On the line `db_names = config["db_name"].split(",")` specifically: `config["db_name"]` is no longer a comma separated string, but already a list so the split can be omitted.

PieterPaulussen commented:
@tishmen An additional refactoring is necessary [here](https://github.com/tishmen/queue/blob/pr/queue_job_base/queue_job/models/queue_job.py#L225) because `read_group` is no longer supported by the core. You'll have to refactor this into the use of the new `_read_group` method.

nilshamerlinck commented:
@PieterPaulussen fyi, workers mode working fine here, is the runner starting properly on your end?

`$ egrep "(runner|runner\.channels): .*(starting|initializing|.+runner ready|database connections ready|Configured channel)" odoo.log`

<img width="1349" height="134" alt="image" src="https://github.com/user-attachments/assets/389288e8-aa3b-421d-93e9-4ea3d33b9915" />

PieterPaulussen commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,43 @@ class QueueJob(models.Model):
-        }
+        uuids = [uuid for uuid in self.mapped("graph_uuid") if uuid]
+        ids_per_graph_uuid = {}
+        if uuids:
```
```suggestion
        ids_per_graph_uuid = {}
        if any(self.mapped("graph_uuid")):
```
Isn't `any` faster in this case?

tishmen replied:
Thanks for the suggestion. Here we still need the full list of UUIDs to build the domain for _read_group, so using any(self.mapped("graph_uuid")) would construct the list and then we’d have to build it again doing two passes instead of one. Since mapped already returns a list, checking if uuids is O(1) and avoids the extra traversal.

PieterPaulussen replied:
~~But 'any' is designed to stop iteration when the first match occurs...~~

PieterPaulussen commented on queue_job/tests/test_json_field.py:
```diff
@@ -14,8 +14,24 @@
+                "login": "queue_demo_user",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
```
This should be replaced by `Command.set()` to be consistent with the rest of your PR.

tishmen replied:
Handled in: https://github.com/OCA/queue/pull/840/commits/c30bf4b55fa841235870c81a842dd55cd28dc711

PieterPaulussen commented on queue_job/tests/test_json_field.py:
```diff
@@ -14,8 +14,24 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
```
```suggestion
        main_company = cls.env.ref("base.main_company")
        group_user = cls.env.ref("base.group_user")
        cls.demo_user = cls.env["res.users"].create(
```
A bit opnionated, but the `User` variable is unnecessary and Camel cased variables should not be used.

sbidoul replied:
Camelcase is ok in this case as `cls.env["res.users"]` is (roughly) a class.

PieterPaulussen commented on queue_job/tests/test_json_field.py:
```diff
@@ -100,7 +116,7 @@ def test_decoder_recordset_list(self):
     def test_decoder_recordset_list_without_user(self):
         value_json = (
-            '["a", 1, {"_type": "odoo_recordset",' '"model": "res.users", "ids": [1]}]'
+            '["a", 1, {"_type": "odoo_recordset","model": "res.users", "ids": [1]}]'
```
```suggestion
            '["a", 1, {"_type": "odoo_recordset", "model": "res.users", "ids": [1]}]'
```

PieterPaulussen commented on queue_job/tests/test_json_field.py:
```diff
@@ -132,7 +148,7 @@ def test_encoder_date(self):
 
     def test_decoder_date(self):
-        value_json = '["a", 1, {"_type": "date_isoformat",' '"value": "2017-04-19"}]'
+        value_json = '["a", 1, {"_type": "date_isoformat","value": "2017-04-19"}]'
```
```suggestion
        value_json = '["a", 1, {"_type": "date_isoformat", "value": "2017-04-19"}]'
```

tishmen replied:
Applied the suggested JSON formatting tweak. Thanks!

PieterPaulussen commented on queue_job/tests/common.py:
```diff
@@ -298,18 +297,24 @@ def _filtered_enqueued_jobs(self, job_method):
+            f"<{call.method.__self__}>."
+            f"{call.method.__name__}({method_args}) "
+            f"with properties ({props_str})"
         )
```
This seems like an unnecessary refactor. The `if` blocks are helpful to separate the args a kwargs processing.

mostafabarmshory replied:
@PieterPaulussen 

Parmeters arg and kwargs are not None always. So no need if block. So this is ok. I thing you should accept this change.

PieterPaulussen replied:
@mostafabarmshory that's not the point I'm trying to make...

tishmen replied:
Updated `_format_job_call()` to use explicit `if` blocks for args/kwargs while still assembling the final string cleanly. Thanks for the suggestion.

PieterPaulussen commented on test_queue_job/tests/test_job.py:
```diff
@@ -32,6 +32,22 @@
+                "login": "queue_demo_user_3",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
```
Should be replaced by `Command.set()`

mostafabarmshory replied:
@tishmen 
@ sbidoul

Please review and approve this PR. We need this branch, and progress is blocked by non-critical issues.

PieterPaulussen commented on test_queue_job/tests/test_job.py:
```diff
@@ -32,6 +32,22 @@
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
```
```suggestion
        main_company = cls.env.ref("base.main_company")
        group_user = cls.env.ref("base.group_user")
        cls.demo_user = cls.env["res.users"].create(
```

mostafabarmshory replied:
@PieterPaulussen 

I don’t think this kind of issue matters. We should accept and publish the PR as long as all test cases pass.

So I big you to accept and merge the RP.

PieterPaulussen replied:
@mostafabarmshory What a weird thing to say. Just because tests are green doesn't make it a valid PR. Unless you have something valuable to add, please wait for the review process to resolve itself.

tishmen replied:
Adjusted the test setup to inline `cls.env["res.users"].create(...)` as suggested. Thanks!

PieterPaulussen commented on test_queue_job/tests/test_job.py:
```diff
@@ -387,6 +403,22 @@ def test_job_identity_key_func_exact(self):
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
```
```suggestion
        main_company = cls.env.ref("base.main_company")
        group_user = cls.env.ref("base.group_user")
        cls.demo_user = cls.env["res.users"].create(
```

tishmen replied:
Same update applied here as well (inline create, removed alias). Thanks!

PieterPaulussen commented on test_queue_job/tests/test_job.py:
```diff
@@ -387,6 +403,22 @@ def test_job_identity_key_func_exact(self):
+                "login": "queue_demo_user_4",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
```
Should be replaced by `Command.set()`

tishmen replied:
Handled in: https://github.com/OCA/queue/pull/840/commits/c30bf4b55fa841235870c81a842dd55cd28dc711

PieterPaulussen commented on test_queue_job/tests/test_job.py:
```diff
@@ -511,6 +543,22 @@ def test_store_env_su_sudo(self):
+                "login": "queue_demo_user_5",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
```
Should be replaced by `Command.set()`

tishmen replied:
Handled in: https://github.com/OCA/queue/pull/840/commits/c30bf4b55fa841235870c81a842dd55cd28dc711

PieterPaulussen commented on test_queue_job/tests/test_json_field.py:
```diff
@@ -11,17 +11,33 @@
+                "login": "queue_demo_user_2",
+                "company_id": main_company.id,
+                "company_ids": [(6, 0, [main_company.id])],
+                "group_ids": [(6, 0, [group_user.id])],
```
Should be replaced by `Command.set()`

tishmen replied:
Handled in: https://github.com/OCA/queue/pull/840/commits/c30bf4b55fa841235870c81a842dd55cd28dc711

PieterPaulussen commented on test_queue_job/tests/test_job.py:
```diff
@@ -511,6 +543,22 @@ def test_store_env_su_sudo(self):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
```
```suggestion
        main_company = cls.env.ref("base.main_company")
        group_user = cls.env.ref("base.group_user")
        cls.demo_user = cls.env["res.users"].create(
```

tishmen replied:
Thanks for the suggestions. In this test suite we deliberately use User = cls.env["res.users"], and @sbidoul  confirmed CamelCase is acceptable here since cls.env["res.users"] behaves like a class. To stay consistent across tests, I’ll keep the current pattern.

PieterPaulussen replied:
@tishmen Please mark the other comments related to this as resolved.

PieterPaulussen commented on test_queue_job/tests/test_json_field.py:
```diff
@@ -11,17 +11,33 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        User = cls.env["res.users"]
+        main_company = cls.env.ref("base.main_company")
+        group_user = cls.env.ref("base.group_user")
+        cls.demo_user = User.create(
```
```suggestion
        main_company = cls.env.ref("base.main_company")
        group_user = cls.env.ref("base.group_user")
        cls.demo_user = cls.env["res.users"].create(
```

tishmen replied:
See above.

PieterPaulussen reviewed (changes requested):
@tishmen Thanks for your effort on this. On the whole, this PR does the trick, but there are a lot of additional changes that are not essential to the module migration. In my opinion, you will get faster approvals if you slim down the changes to the bare minimum given the complexity of this module.

If you want to keep the translations in this PR, can we go one step further and use positional arguments for all translated string variables? This results in more flexibility for other languages during translating.

On the refactoring of the tests to remove the reliance on the demo_user, I would beg to differ. Unittests should rely on the presence of the demo data, just as the core does.

Finally, I would like to stress the following issue that I encountered during a local test with your code: https://github.com/OCA/queue/pull/840#issuecomment-3449954152
The jobrunner does not seem to enqueue AND process the queue jobs when running is preforked server configuration. This issue should be addressed first.

PieterPaulussen commented:
> @PieterPaulussen fyi, workers mode working fine here, is the runner starting properly on your end?
> 
> `$ egrep "(runner|runner\.channels): .*(starting|initializing|.+runner ready|database connections ready|Configured channel)" odoo.log`
> 
> <img alt="image" width="1349" height="134" src="https://private-user-images.githubusercontent.com/511258/506441972-389288e8-aa3b-421d-93e9-4ea3d33b9915.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE2OTI2OTIsIm5iZiI6MTc2MTY5MjM5MiwicGF0aCI6Ii81MTEyNTgvNTA2NDQxOTcyLTM4OTI4OGU4LWFhM2ItNDIxZC05M2U5LTRlYTNkMzNiOTkxNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyOFQyMjU5NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYzhhOTMzZjZkMTEwNjc2Y2I2ZTBiYjY0NWJjNTllM2M2MjdlMzY5MmE3OTA3NGJiM2Q0MDkwNzEzMjAxYWIyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bCQ2xH_iSKnEVhrAcz2qm4R2a5SuwlAEavJH9k-fQHk">

@nilshamerlinck Curious, could you share your worker and channel configuration please? In my test, the jobs got enqueued, but were never actually processed.

maciej-wichowski commented:
@tishmen Please take a look at [this commit](https://github.com/versada/queue/commit/d3d8f2cef2785f7fd608b8c00eb7548504d72be7). I would create a PR for your branch, but for some reason I couldn't do it.

It's a fix related to a change in `db_name` parsing from config ([src](https://github.com/odoo/odoo/commit/5710ba712c3c50a74a816f1270d159b62f3a1b95#diff-dcb8729ab89f3d99d7980d482a99df7d3ecb1c82e14fad9c3778d2606b97a120))

CasVissers-360ERP commented:
@tishmen 
 jobs_groups = self.env["queue.job"].read_group(
 
 I get: DeprecationWarning: Since 19.0, read_group is deprecated. Please use _read_group might be good to get rid of this?

omarsyscoon commented:
@tishmen 
i have this when i initialise new db,

```
DR-ODOO  | 2025-11-04 12:13:16,520 1 DEBUG ? odoo.addons.queue_job.jobrunner.runner: initializing database connections 
DR-ODOO  | 2025-11-04 12:13:16,520 1 ERROR ? odoo.addons.queue_job.jobrunner.runner: exception: sleeping 5s and retrying 
DR-ODOO  | Traceback (most recent call last):
DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 597, in run
DR-ODOO  |     self.initialize_databases()
DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 492, in initialize_databases
DR-ODOO  |     for db_name in sorted(self.get_db_names()):
DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 476, in get_db_names
DR-ODOO  |     db_names = config["db_name"].split(",")
DR-ODOO  | AttributeError: 'list' object has no attribute 'split'
DR-ODOO  | 2025-11-04 12:13:17,230 1 INFO 19_controller3 odoo.modules.loading: loading base/data/report_paperformat_data.xml 

```

CasVissers-360ERP commented:
> @tishmen i have this when i initialise new db,
> 
> ```
> DR-ODOO  | 2025-11-04 12:13:16,520 1 DEBUG ? odoo.addons.queue_job.jobrunner.runner: initializing database connections 
> DR-ODOO  | 2025-11-04 12:13:16,520 1 ERROR ? odoo.addons.queue_job.jobrunner.runner: exception: sleeping 5s and retrying 
> DR-ODOO  | Traceback (most recent call last):
> DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 597, in run
> DR-ODOO  |     self.initialize_databases()
> DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 492, in initialize_databases
> DR-ODOO  |     for db_name in sorted(self.get_db_names()):
> DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 476, in get_db_names
> DR-ODOO  |     db_names = config["db_name"].split(",")
> DR-ODOO  | AttributeError: 'list' object has no attribute 'split'
> DR-ODOO  | 2025-11-04 12:13:17,230 1 INFO 19_controller3 odoo.modules.loading: loading base/data/report_paperformat_data.xml 
> ```

I have this issue as well, are you on Odoo.sh?

omarsyscoon commented:
> > @tishmen i have this when i initialise new db,
> > ```
> > DR-ODOO  | 2025-11-04 12:13:16,520 1 DEBUG ? odoo.addons.queue_job.jobrunner.runner: initializing database connections 
> > DR-ODOO  | 2025-11-04 12:13:16,520 1 ERROR ? odoo.addons.queue_job.jobrunner.runner: exception: sleeping 5s and retrying 
> > DR-ODOO  | Traceback (most recent call last):
> > DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 597, in run
> > DR-ODOO  |     self.initialize_databases()
> > DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 492, in initialize_databases
> > DR-ODOO  |     for db_name in sorted(self.get_db_names()):
> > DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 476, in get_db_names
> > DR-ODOO  |     db_names = config["db_name"].split(",")
> > DR-ODOO  | AttributeError: 'list' object has no attribute 'split'
> > DR-ODOO  | 2025-11-04 12:13:17,230 1 INFO 19_controller3 odoo.modules.loading: loading base/data/report_paperformat_data.xml 
> > ```
> 
> I have this issue as well, are you on Odoo.sh?

no i was testing locally, 

temporary i solved by the following 
jobrunner/runner.py L#474
```

    def get_db_names(self):
        if config["db_name"]:
            if isinstance(config["db_name"], list):
                return config["db_name"]
            return config["db_name"].split(",")
        return odoo.service.db.list_dbs(True)

```

Reyes4711-S73 reviewed (approved):
LGTM

omarsyscoon reviewed (approved):
LGTM

tishmen commented:
> @tishmen Please take a look at [this commit](https://github.com/versada/queue/commit/d3d8f2cef2785f7fd608b8c00eb7548504d72be7). I would create a PR for your branch, but for some reason I couldn't do it.
> 
> It's a fix related to a change in `db_name` parsing from config ([src](https://github.com/odoo/odoo/commit/5710ba712c3c50a74a816f1270d159b62f3a1b95#diff-dcb8729ab89f3d99d7980d482a99df7d3ecb1c82e14fad9c3778d2606b97a120))

Handled in update.

tishmen commented:
> @tishmen jobs_groups = self.env["queue.job"].read_group(
> 
> I get: DeprecationWarning: Since 19.0, read_group is deprecated. Please use _read_group might be good to get rid of this?

Handled in update.

tishmen commented:
> @tishmen i have this when i initialise new db,
> 
> ```
> DR-ODOO  | 2025-11-04 12:13:16,520 1 DEBUG ? odoo.addons.queue_job.jobrunner.runner: initializing database connections 
> DR-ODOO  | 2025-11-04 12:13:16,520 1 ERROR ? odoo.addons.queue_job.jobrunner.runner: exception: sleeping 5s and retrying 
> DR-ODOO  | Traceback (most recent call last):
> DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 597, in run
> DR-ODOO  |     self.initialize_databases()
> DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 492, in initialize_databases
> DR-ODOO  |     for db_name in sorted(self.get_db_names()):
> DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 476, in get_db_names
> DR-ODOO  |     db_names = config["db_name"].split(",")
> DR-ODOO  | AttributeError: 'list' object has no attribute 'split'
> DR-ODOO  | 2025-11-04 12:13:17,230 1 INFO 19_controller3 odoo.modules.loading: loading base/data/report_paperformat_data.xml 
> ```

Handled in update.

tishmen commented:
> > > @tishmen i have this when i initialise new db,
> > > ```
> > > DR-ODOO  | 2025-11-04 12:13:16,520 1 DEBUG ? odoo.addons.queue_job.jobrunner.runner: initializing database connections 
> > > DR-ODOO  | 2025-11-04 12:13:16,520 1 ERROR ? odoo.addons.queue_job.jobrunner.runner: exception: sleeping 5s and retrying 
> > > DR-ODOO  | Traceback (most recent call last):
> > > DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 597, in run
> > > DR-ODOO  |     self.initialize_databases()
> > > DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 492, in initialize_databases
> > > DR-ODOO  |     for db_name in sorted(self.get_db_names()):
> > > DR-ODOO  |   File "/mnt/odoo/external_addons/queue_job/jobrunner/runner.py", line 476, in get_db_names
> > > DR-ODOO  |     db_names = config["db_name"].split(",")
> > > DR-ODOO  | AttributeError: 'list' object has no attribute 'split'
> > > DR-ODOO  | 2025-11-04 12:13:17,230 1 INFO 19_controller3 odoo.modules.loading: loading base/data/report_paperformat_data.xml 
> > > ```
> > 
> > 
> > I have this issue as well, are you on Odoo.sh?
> 
> no i was testing locally,
> 
> temporary i solved by the following jobrunner/runner.py L#474
> 
> ```
> 
>     def get_db_names(self):
>         if config["db_name"]:
>             if isinstance(config["db_name"], list):
>                 return config["db_name"]
>             return config["db_name"].split(",")
>         return odoo.service.db.list_dbs(True)
> ```

I am going to use a similar solution to yours to handle this and not the PR that was suggested above. Thanks.

tishmen commented:
> @tishmen Thanks for your effort on this. On the whole, this PR does the trick, but there are a lot of additional changes that are not essential to the module migration. In my opinion, you will get faster approvals if you slim down the changes to the bare minimum given the complexity of this module.
> 
> If you want to keep the translations in this PR, can we go one step further and use positional arguments for all translated string variables? This results in more flexibility for other languages during translating.
> 
> On the refactoring of the tests to remove the reliance on the demo_user, I would beg to differ. Unittests should rely on the presence of the demo data, just as the core does.
> 
> Finally, I would like to stress the following issue that I encountered during a local test with your code: [#840 (comment)](https://github.com/OCA/queue/pull/840#issuecomment-3449954152) The jobrunner does not seem to enqueue AND process the queue jobs when running is preforked server configuration. This issue should be addressed first.

Thanks for the review. I’ve updated the PR to use positional arguments for all translated strings to improve i18n flexibility. Regarding the tests: in Odoo 19, demo data isn’t loaded by default during test runs, which caused failures when relying on the demo user, so the tests create a user explicitly to stay deterministic

aisopuro commented:
For extra context, Odoo indeed changed the default for loading demo data in https://github.com/odoo/odoo/commit/eeac3d43a645bf7bb31b38c951d860163de7a458.

They even explicitly state in the commit message:
> In testing we should not rely on demo data, but instead data constructed during the test.

I at least would interpret this to mean that part of migrating code to 19 is to rewrite/refactor tests so they no longer depend on demo data.

mostafabarmshory reviewed (approved):
@tishmen 
@sbidoul 
@PieterPaulussen 

Please review and approve this PR. We need this branch, and progress is blocked by non-critical issues.

PieterPaulussen reviewed (approved):
Most of my feedback seems to be either ignored or resolved.

Robrechtc commented:
@pedrobaeza Sorry for tagging you here again but you are the only one that I know has merge rights. Would you mind merging this? Thanks in advance.

pedrobaeza commented:
Hi, I'm afraid I don't know too much this module, so one PSC like @guewen, @simahawk or @sbidoul should check that it's correct and then merge.

Robrechtc commented:
> Hi, I'm afraid I don't know too much this module, so one PSC like @guewen, @simahawk or @sbidoul should check that it's correct and then merge.

Thanks for tagging them.

sbidoul commented:
I plan to read the latest comments and re-review soon(ish). Maybe next week.

It the meantime, nothing prevents using it, and reporting further issues if any.

richard-willdooit commented:
@tishmen Well done - all I can really say is, that I'm glad my PR was superseded - :stuck_out_tongue_winking_eye:  

Overall, I really think we should be getting these PRs for migrations through as efficiently as possible.  I know everyone has their opinions, and where code has been touched or altered, it is great to aim for perfection, but if the changes are not wrong, if they are not regressive, then there should be no obligation to "improve" during a migration.  Nice if it does, but it should not be obligatory.

After watching this unfold, I will be VERY uninclined to offer a module migration, as my objective will be to get the module across as is, and correct, not suddenly inheriting everyone else's wishlists...

Of course, this does not mean dismissing anyone's reviews or input without due and proper consideration, but the end game of a migration PR seems different for different people....

sbidoul commented on queue_job/jobrunner/runner.py:
```diff
@@ -472,11 +472,12 @@ def from_environ_or_config(cls):
+        if db_name_opt:
+            if isinstance(db_name_opt, (list, tuple, set)):
+                return list(db_name_opt)
+            return [n for n in str(db_name_opt).split(",") if n]
```
Does any one know in which case `config["db_name"]` is not a list? As I read `odoo/tools/config.py` it seems it's always a list, since it is parsed with `_check_comma`.

tishmen replied:
Good point—simplified `get_db_names()` to treat `config["db_name"]` as the list parsed by Odoo (`type='comma'`) and removed the CSV fallback. Thanks!

sbidoul commented on queue_job/models/base.py:
```diff
@@ -260,11 +269,5 @@ def _job_prepare_context_before_enqueue(self):
-        wrapped.origin = origin
-        setattr(cls, name, wrapped)
+    # Note: no local _patch_method helper; if needed, patch methods
+    # directly in _register_hook as shown above.
```
```suggestion
```

Nit: this comment is not really helpful outside the migration PR.

tishmen replied:
Removed the migration-only comment. Thanks!

sbidoul commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,43 @@ class QueueJob(models.Model):
     def init(self):
+        cr = self.env.cr
         index_1 = "queue_job_identity_key_state_partial_index"
         index_2 = "queue_job_channel_date_done_date_created_index"
```
nit: these two variables are not helpful, they could be inlined

tishmen replied:
Inlined the index name strings (removed the temporary variables). Thanks!

sbidoul commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,43 @@ class QueueJob(models.Model):
-        ids_per_graph_uuid = {
-            group["graph_uuid"]: group["ids"] for group in jobs_groups
-        }
+        uuids = [uuid for uuid in self.mapped("graph_uuid") if uuid]
```
```suggestion
        graph_uuids = [uuid for uuid in self.mapped("graph_uuid") if uuid]
```

To avoid confusion with job uuids.

tishmen replied:
Renamed `uuids` to `graph_uuids` to avoid confusion with job UUIDs. Thanks!

sbidoul commented on queue_job/models/queue_job.py:
```diff
@@ -127,38 +128,43 @@ class QueueJob(models.Model):
+            )
+            # rows -> list of tuples: (graph_uuid, recordset)
+            for graph_uuid, recs in rows:
+                ids_per_graph_uuid[graph_uuid] = recs.ids
```
```suggestion
            ids_per_graph_uuid = dict(self.env["queue.job"]._read_group(
                [("graph_uuid", "in", graph_uuids)],
                groupby=["graph_uuid"],
                aggregates=["id:array_agg"],
            ))
```

This is simpler and more efficient, as we don't need the recordsets.

tishmen replied:
Switched to `_read_group(..., aggregates=["id:array_agg"])` and `dict(...)` to avoid building recordsets just to get ids. Thanks!

sbidoul commented on queue_job/models/queue_job.py:
```diff
@@ -216,20 +222,16 @@ def _dependency_graph_vis_node(self):
+                ["graph_uuid"],
+                ["__count"],
+            )
+            count_per_graph_uuid = {graph_uuid: cnt for graph_uuid, cnt in rows}
```
nit, this is probably slightly more efficient than the dict comprehension

```suggestion
            count_per_graph_uuid = dict(self.env["queue.job"]._read_group(
                [("graph_uuid", "in", graph_uuids)],
                groupby=["graph_uuid"],
                aggregates=["__count"],
            ))
```

tishmen replied:
Applied the `dict(_read_group(...))` tweak here as well. Thanks!

sbidoul reviewed (commented):
Thanks, this looks great now.

A few last minor things, and one more important one for performance of graph handling in the read_group rewrite.

tishmen commented:
@sbidoul I will take care of all of the code review comments this weekend. Thanks for your feedback.

em230418 commented:
Здраво, @tishmen!

> @sbidoul I will take care of all of the code review comments this weekend. 

Hope you had nice weekend :D

tishmen commented:
Sorry for the delay — I’ve pushed updates addressing the review feedback (style nits, test cleanups, and the _read_group perf tweak). Pre-commit is green again. Please take another look when you have a moment.

sbidoul commented on base_import_async/models/base_import_import.py:
```diff
@@ -55,10 +55,9 @@ def execute_import(self, fields, columns, options, dryrun=False):
-        }
+        description = _(
+            "Import %s from file %s", translated_model_name, self.file_name
+        )
```
```suggestion
        description = _(
            "Import %(model)s from file %(from_file)s", model=translated_model_name, from_file=self.file_name
        )
```

It's important to 
1/ keep the source string unchanged
2/ use keyword placeholders so it can be translated to languages that need another order of the words

sbidoul commented on queue_job/jobrunner/runner.py:
```diff
@@ -472,11 +472,10 @@ def from_environ_or_config(cls):
-        return db_names
+        db_names = config["db_name"] or []
+        if db_names:
+            return list(db_names)
```
```suggestion
            return db_names
```

It's already a list, no?

sbidoul commented on base_import_async/models/base_import_import.py:
```diff
@@ -156,16 +155,13 @@ def _split_file(
+                file_name,
+                chunk,
+                row_from + 1 + header_offset,
+                row_to + 1 + header_offset,
```
Same here, please don't change the source string.

etobella commented:
@sbidoul can we help somehow on this?

Just a question, there is need for so many migration commits or should it be squashed in a single one?

sbidoul commented on queue_job/jobrunner/runner.py:
```diff
@@ -472,11 +472,10 @@ def from_environ_or_config(cls):
-        else:
-            db_names = odoo.service.db.list_dbs(True)
-        return db_names
+        db_names = config["db_name"] or []
```
```suggestion
        db_names = config["db_name"]
```

It's superfluous.

sbidoul reviewed (commented):
I had these last comments pending.

sbidoul commented:
About the commits, this helped during review but I suppose it's best to squash them now. Then we can merge.

milan-dd committed:
[MIG] queue_job: migrate + tests

- Migrate queue_job, test_queue_job and base_import_async to 19.0

tishmen commented:
Hi all! I’ve now squashed PR #840 down to a single commit as requested, and kept the latest review fixes in place. Pre-commit and an Odoo --stop-after-init install check are green on my side. Could you please re-review when you have a moment? @sbidoul @etobella

sbidoul commented:
Thanks for all this work @tishmen !

/ocabot merge nobump

tishmen commented:
@sbidoul Thanks! And also huge thanks to everyone who reviewed, commented, and helped move this forward 🙏

sbidoul commented:
See also #872 for the latest improvements from 18. Also we should replace the autovacuum cron with a `@api.autovacuum` decorator (with a migration to remove the cron). If anyone feels like doing this :)

tishmen commented:
Thanks for the pointer! 👍@sbidoul

I’ll take care of this in a separate PR this weekend (switching to the @api.autovacuum decorator and adding the migration to remove the cron).

Also, it would be greatly appreciated if the remaining OCA Queue–related PRs could get some review time as well. I really like to get them ready for merging.

📣 Call for help: if anyone reading this has some bandwidth to review or test the other Queue PRs, help would be very welcome. Thanks everyone!

PR Links:

https://github.com/OCA/queue/pull/841
https://github.com/OCA/queue/pull/842
https://github.com/OCA/queue/pull/843
https://github.com/OCA/queue/pull/844
https://github.com/OCA/queue/pull/845
